src/HealthGPS/risk_factor_adjustable_model.cpp

```diff
@@ -0,0 +1,170 @@
+#include "HealthGPS.Core/thread_util.h"
+
+#include "risk_factor_adjustable_model.h"
+#include "sync_message.h"
+
+namespace { // anonymous namespace
+
+/// @brief Defines the risk factors adjustment synchronisation message
+using RiskFactorAdjustmentMessage = hgps::SyncDataMessage<hgps::RiskFactorSexAgeTable>;
+
+/// @brief Defines the first statistical moment type
+struct FirstMoment {
+  public:
+    void append(double value) noexcept {
+        count_++;
+        sum_ += value;
+    }
+
+    double mean() const noexcept { return (count_ > 0) ? (sum_ / count_) : 0.0; }
+
+  private:
+    int count_{};
+    double sum_{};
+};
+
+} // anonymous namespace
+
+namespace hgps {
+
+RiskFactorAdjustableModel::RiskFactorAdjustableModel(
+    const RiskFactorSexAgeTable &risk_factor_expected)
+    : risk_factor_expected_{risk_factor_expected} {
+
+    if (risk_factor_expected_.empty()) {
+        throw core::HgpsException("Risk factor expected value mapping is empty");
+    }
+}
+
+const RiskFactorSexAgeTable &RiskFactorAdjustableModel::get_risk_factor_expected() const noexcept {
+    return risk_factor_expected_;
+}
+
+void RiskFactorAdjustableModel::adjust_risk_factors(
+    RuntimeContext &context, const std::vector<core::Identifier> &keys) const {
+    RiskFactorSexAgeTable adjustments;
+
+    // Baseline scenatio: compute adjustments.
+    if (context.scenario().type() == ScenarioType::baseline) {
+        adjustments = calculate_adjustments(context, keys);
+    }
+
+    // Intervention scenario: recieve adjustments from baseline scenario.
+    else {
+        auto message = context.scenario().channel().try_receive(context.sync_timeout_millis());
+        if (!message.has_value()) {
+            throw core::HgpsException(
+                "Simulation out of sync, receive baseline adjustments message has timed out");
+        }
+
+        auto &basePtr = message.value();
+        auto *messagePrt = dynamic_cast<RiskFactorAdjustmentMessage *>(basePtr.get());
+        if (!messagePrt) {
+            throw core::HgpsException(
+                "Simulation out of sync, failed to receive a baseline adjustments message");
+        }
+
+        adjustments = messagePrt->data();
+    }
+
+    // All scenarios: apply adjustments to population.
+    auto &pop = context.population();
+    std::for_each(core::execution_policy, pop.begin(), pop.end(), [&](auto &person) {
+        if (!person.is_active()) {
+            return;
+        }
+
+        for (const auto &factor : keys) {
+            double delta = adjustments.at(person.gender, factor).at(person.age);
+            person.risk_factors.at(factor) += delta;
+        }
+    });
+
+    // Baseline scenario: send adjustments to intervention scenario.
+    if (context.scenario().type() == ScenarioType::baseline) {
+        context.scenario().channel().send(std::make_unique<RiskFactorAdjustmentMessage>(
+            context.current_run(), context.time_now(), std::move(adjustments)));
+    }
```

Comment:
If this is sending the adjustments to the intervention scenario, is this not some sort of recurrent function that runs once the baseline and then, immediately afterwards, the intervention cases, but doing different things in both?

Maybe I'm missing something but this feels as it should be 3 functions:

1. Apply adjustments, whatever their origin
2. Apply adjustments from intervention scenario (calling 1 inside)
3. Apply adjustments from baseline scenario (calling `calculate_adjustments` and 2 inside).

As I said, I might be missing the actual workflow of the application. I'm not totally sure when messages are executed. If sending a message using `pypubsub` in python, things are purely sequential and syncronous, but I don't know if this works the same way.

Reply:
I'm not completely sure I understand the question, but the pipe is unidirectional; baseline -> intervention. Does that answer?

Reply:
No quite. My point is that this function does too many things depending on what the scenario is and, it seems to me, it also calls itself to go from baseline to intervention. I'm not sure how the messaging system work in C++, but it feels this function should be split into multiple ones doing one thing only each (and possibly calling each other).

Reply:
But not a barrier for this PR.

Reply:
Will make an issue and think about it.